src/JSONAPIExtension.php

```diff
@@ -63,11 +65,41 @@ public static function getSubscribedEvents()
             KernelEvents::EXCEPTION => [
                 ['error', 515],
             ],
+            KernelEvents::CONTROLLER => [
+                ['disableFrontend', -1025]
+            ]
         ];
 
         return $parentEvents + $localEvents;
     }
 
+    public function disableFrontend(FilterControllerEvent $event)
+    {
+        $request = $event->getRequest();
+        //$response = $event->getResponse();
+
+        $container = $this->getContainer();
+
+        //Get route name
+        $routeName = $request->get('_route');
+
+        //Check if request is to frontend
+        if (Zone::isFrontend($request)) {
+            //Check if we should disable frontend based upon the configuration
+            if ($container['jsonapi.config']->isDisableFrontend()) {
+                //Only disable frontend routes, don't disable json routes
+                if (strpos($routeName, 'jsonapi') === false) {
+                    $event->stopPropagation();
+                    $event->setController(
+                        function() {
+                            return new Response('', 200);
+                        }
+                    );
+                }
+            }
+        }
```

Comment:
Can you un-nest this. Early returns are far better for human reading of logical intent, and there is rarely a time where an if statement inside another is valid.